File: nakari/search/providers.py
"""Search provider implementations."""

import httpx
from typing import Any

from .errors import SearchAuthError, SearchError, SearchRateLimitError
from .types import SearchOptions, SearchResponse, SearchResponse as SearchResult, SearchType


class SearchProvider:
    """Base interface for search providers."""

    @property
    def name(self) -> str:
        """Provider name."""
        raise NotImplementedError

    async def search(
        self, query: str, options: SearchOptions | None = None
    ) -> SearchResponse:
        """Execute a search query.

        Args:
            query: Search query string
            options: Optional search parameters

        Returns:
            SearchResponse with results

        Raises:
            SearchError: If the search fails
        """
        raise NotImplementedError


class SerperProvider(SearchProvider):
    """Serper.dev search provider implementation.

    Serper provides Google Search results via REST API.
    """

    DEFAULT_BASE_URL = "https://google.serper.dev"

    def __init__(
        self,
        api_key: str,
        base_url: str | None = None,
        timeout: int = 20,
    ) -> None:
        """Initialize the Serper provider.

        Args:
            api_key: Serper API key
            base_url: Optional base URL (defaults to https://google.serper.dev)
            timeout: Request timeout in seconds
        """
        self._api_key = api_key
        self._base_url = base_url or self.DEFAULT_BASE_URL
        self._timeout = timeout
        self._client = httpx.AsyncClient(timeout=timeout)

    @property
    def name(self) -> str:
        """Provider name."""
        return "serper"

    async def search(
        self, query: str, options: SearchOptions | None = None
    ) -> SearchResponse:
        """Execute a search query.

        Args:
            query: Search query string
            options: Optional search parameters

        Returns:
            SearchResponse with results

        Raises:
            SearchAuthError: If authentication fails
            SearchRateLimitError: If rate limit is exceeded
            SearchError: For other errors
        """
        opts = options or SearchOptions()

        # Determine endpoint based on search type
        endpoints: dict[SearchType, str] = {
            "search": "/search",
            "news": "/news",
            "images": "/images",
            "videos": "/videos",
        }
        endpoint = endpoints.get(opts.type, "/search")

        # Build request payload
        payload: dict[str, Any] = {"q": query}
        if opts.num_results:
            payload["num"] = opts.num_results
        if opts.language:
            payload["hl"] = opts.language
        if opts.region:
            payload["gl"] = opts.region

        headers = {
            "X-API-KEY": self._api_key,
            "Content-Type": "application/json",
        }

        url = f"{self._base_url}{endpoint}"

        try:
            response = await self._client.post(url, json=payload, headers=headers)

            # Handle error responses
            if response.status_code in (401, 403):
                raise SearchAuthError(
                    f"Authentication failed: {response.text}"
                ).with_context(status_code=response.status_code)
            elif response.status_code == 429:
                raise SearchRateLimitError(
                    f"Rate limit exceeded: {response.text}"
                ).with_context(status_code=response.status_code)
            elif response.status_code >= 400:
                raise SearchError(
                    f"Search failed: {response.text}",
                    code="HTTP_ERROR",
                ).with_context(status_code=response.status_code)

            data = response.json()

            # Convert to SearchResponse
            results = self._convert_response(data, opts.type)
            return SearchResponse(
                results=results,
                total_results=data.get("searchInformation", {}).get(
                    "totalResults"
                ),
                query=query,
            )

        except httpx.TimeoutException:
            raise SearchError("Request timed out", code="TIMEOUT_ERROR").with_context(
                timeout=self._timeout
            )
        except httpx.RequestError as e:
            raise SearchError(
                f"Request failed: {e!s}", code="REQUEST_ERROR"
            ).with_context(error=str(e))
        except (SearchAuthError, SearchRateLimitError, SearchError):
            raise
        except Exception as e:
            raise SearchError(
                f"Unexpected error: {e!s}", code="UNKNOWN_ERROR"
            ).with_context(error=str(e))

    def _convert_response(
        self, data: dict[str, Any], search_type: SearchType
    ) -> list[dict[str, Any]]:
        """Convert Serper API response to standardized results.

        Args:
            data: Raw API response data
            search_type: Type of search performed

        Returns:
            List of standardized result dictionaries
        """
        results: list[dict[str, Any]] = []

        # Different result keys based on search type
        result_keys: dict[SearchType, str] = {
            "search": "organic",
            "news": "news",
            "images": "images",
            "videos": "videos",
        }
        key = result_keys.get(search_type, "organic")
        raw_results = data.get(key, [])

        for item in raw_results:
            result: dict[str, Any] = {
                "title": item.get("title", ""),
                "link": item.get("link", ""),
                "snippet": item.get("snippet", ""),
            }

            # Add type-specific fields
            if search_type == "news":
                result["date"] = item.get("date")
                result["siteName"] = item.get("source")
            elif search_type == "images":
                result["thumbnailUrl"] = item.get("imageUrl")
            elif search_type == "videos":
                result["thumbnailUrl"] = item.get("thumbnail")

            results.append(result)

        return results

    async def close(self) -> None:
        """Close the HTTP client."""
        await self._client.aclose()

File: src/search/providers.ts
import type { SearchResponse, SearchOptions, SearchResult } from "./types.js";
import { SearchError, SearchAuthError, SearchRateLimitError } from "./errors.js";

/**
 * Abstract interface for search providers.
 * All search implementations must implement this interface.
 */
export interface SearchProvider {
  /**
   * Human-readable provider name.
   */
  readonly name: string;

  /**
   * Execute a search query.
   *
   * @param query - The search query string
   * @param options - Optional search parameters
   * @returns Standardized search response
   * @throws {SearchError} On provider errors or network failures
   */
  search(query: string, options?: SearchOptions): Promise<SearchResponse>;
}

/**
 * Configuration for Serper.dev provider.
 */
export interface SerperConfig {
  /** API key from https://serper.dev */
  apiKey: string;
  /** Base URL (default: https://google.serper.dev) */
  baseUrl?: string;
  /** Default timeout in milliseconds */
  timeout?: number;
}

/**
 * Serper.dev search provider implementation.
 *
 * Serper.dev provides Google Search results via a simple REST API.
 * Documentation: https://serper.dev/api-guide
 */
export class SerperProvider implements SearchProvider {
  readonly name = "Serper.dev";
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly defaultTimeout: number;

  constructor(config: SerperConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.baseUrl ?? "https://google.serper.dev";
    this.defaultTimeout = config.timeout ?? 10000;
  }

  async search(query: string, options?: SearchOptions): Promise<SearchResponse> {
    const timeout = options?.timeout ?? this.defaultTimeout;
    const numResults = options?.numResults ?? 10;

    // Build request body based on Serper.dev API
    const body: Record<string, unknown> = {
      q: query,
      num: numResults,
    };

    // Add type-specific parameters
    if (options?.type === "news") {
      body.type = "news";
    } else if (options?.type === "images") {
      body.type = "images";
    } else if (options?.type === "videos") {
      body.type = "videos";
    }

    if (options?.language) {
      body.hl = options.language;
    }

    if (options?.region) {
      body.gl = options.region;
    }

    // Create abort controller for timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const response = await fetch(`${this.baseUrl}/search`, {
        method: "POST",
        headers: {
          "X-API-KEY": this.apiKey,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      if (!response.ok) {
        if (response.status === 401 || response.status === 403) {
          throw new SearchAuthError();
        }
        if (response.status === 429) {
          throw new SearchRateLimitError();
        }
        const errorText = await response.text().catch(() => "Unknown error");
        throw new SearchError(`Serper.dev API error: ${response.status} ${response.statusText}`, {
          statusCode: response.status,
          body: errorText,
        });
      }

      const data = (await response.json()) as SerperResponse;
      return this.transformResponse(data, query);
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        throw new SearchError(`Search timeout after ${timeout}ms`);
      }
      if (err instanceof SearchError) {
        throw err;
      }
      throw new SearchError(
        `Failed to execute search: ${err instanceof Error ? err.message : String(err)}`,
      );
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Transform Serper.dev response to standardized format.
   */
  private transformResponse(data: SerperResponse, query: string): SearchResponse {
    const results: SearchResult[] = [];

    // Organic results
    if (data.organic) {
      for (const item of data.organic) {
        results.push({
          title: item.title,
          link: item.link,
          snippet: item.snippet,
          thumbnailUrl: item.thumbnailUrl,
        });
      }
    }

    // News results (if type is "news")
    if (data.news) {
      for (const item of data.news) {
        results.push({
          title: item.title,
          link: item.link,
          snippet: item.snippet,
          date: item.date,
          thumbnailUrl: item.imageUrl,
          siteName: item.source,
        });
      }
    }

    return {
      results,
      totalResults: results.length,
      query,
      metadata: {
        searchParameters: data.searchParameters,
        relatedSearches: data.relatedSearches,
      },
    };
  }
}

/**
 * Serper.dev API response types.
 * See: https://serper.dev/api-guide
 */
interface SerperResponse {
  searchParameters?: Record<string, unknown>;
  organic?: Array<{
    title: string;
    link: string;
    snippet: string;
    thumbnailUrl?: string;
  }>;
  news?: Array<{
    title: string;
    link: string;
    snippet: string;
    date: string;
    source: string;
    imageUrl?: string;
  }>;
  relatedSearches?: Array<{
    query: string;
  }>;
}
